### ISSUE TITLE
Broken PercentageRunnersBusy autoscaling when stalled runner

### ISSUE THREAD

**Issue by alanbover** [actions-runner-controller/actions-runner-controller#1374]
**Describe the bug**
We are currently using PercentageRunnersBusy metric for autoscale the runners. We have found out that sometimes (because an unknown reason) workers are infinitely kept running a workflow.

When this happens and that pod is marked for downscaling (we are using ephemeral runners) the pod is marked with flag AnnotationKeyRunnerCompletionWaitStartTimestamp (here https://github.com/actions-runner-controller/actions-runner-controller/blob/master/controllers/runner_graceful_stop.go#L178). I understand that this happens because Github replies a 422 error code, and hence the runner should be kept registered in Github. It makes sense also that `tickRunnerGracefulStop` kinds of "forgets" about that pod and leaves the reconciler to handle the pod termination event.

What we are seeing though (maybe it's just because some timeout on a heartbeat between the stalled runner and Github) is that the runner is not registered in Github.

This is actually a problem because that pod is not handled then correctly for autoscaling (here https://github.com/actions-runner-controller/actions-runner-controller/blob/master/controllers/autoscaling.go#L312).

As an example, we have 3 runners in the cluster (all bussy) but only 2 are registered in Github. 
```
fractionBusy := float64(2) / float64(3)       // fractionBusy = 0.66 and should be 1
```

I believe that if autoscaling would take into consideration the runners that are not registered in Github this issue would disappear, something like:
```
fractionBusy := (float64(numRunnersBusy) + float64(numRunnersBusyNotRegistered)) / float64(desiredReplicasBefore)         //  fractionBusy = (2 + 1) / (3) = 1
```

We are though currently trying to migrate to Webhook driven scaling, so I don't think we will be able to provide more information (hopefully what it's being reported is enough)

**Checks**

- [x] My actions-runner-controller version (v0.22.0) does support the feature
- [ ] I'm using an unreleased version of the controller I built from HEAD of the default branch

**To Reproduce**
Steps to reproduce the behaviour:
I'm not sure how to reproduce it, as I don't fully understand why the runner gets unregistered from Github if it's still running

**Expected behaviour**
Autoscaler is able to detect that a pod with AnnotationKeyRunnerCompletionWaitStartTimestamp annotation is not registered in Github and it counts it as an extra bussy runner

**Screenshots**

**Environment (please complete the following information):**
 - Controller Version [e.g. 0.22.0]
 - Deployment Method [Helm]
 - Helm Chart Version [0.12.8]

**Additional context**

**Comment by toast-gear** [comment]
Ref https://github.com/actions-runner-controller/actions-runner-controller/issues/1304

Just linking this with issue #1304 as they are related issues

**Comment by mumoshu** [comment]
I realized that the linked issue #1304 had been already addressed in 0.24.0 as explained in https://github.com/actions-runner-controller/actions-runner-controller/issues/1304#issuecomment-1155903896. However, this issue seems like unrelated to it.

**Comment by mumoshu** [comment]
@alanbover Hey! Thanks for reporting. Your suggested change does make sense to me now. I'll address it before 0.25.0

**Comment by mumoshu** [comment]
I was rereading this today and realized that I might have noticed a misconception.

> ```
> fractionBusy := (float64(numRunnersBusy) + float64(numRunnersBusyNotRegistered)) / float64(desiredReplicasBefore)         //  fractionBusy = (2 + 1) / (3) = 1
> ```

You mentioned this but actually what ARC sees is the response of ListRunners API, which includes only registered runners so the updated formula would be more like:


```
> fractionBusy := (float64(numRunnersBusyAccordingToListRunnersAPI) + float64(numRunnersSeeminglyBusy)) / float64(desiredReplicasBefore)         //  fractionBusy = (2 + 1) / (3) = 1
```

The key here is that we currently cache ListRunnersAPI responses for 60 seconds (as that's what GitHub returns for the Cache-Control header), so you might not see `numRunnersBusy`NOT  reflect the recently-turned-busy runners for up to 60 seconds.

I thought it was fine to delay scaling for up to 60 seconds when you're using pull-based scaling hence the current formula. But I take this as a good chance to figure out how we can improve it. Perhaps we can mark the runner pod got 422 as "busy" so that we can count such runners to produce `numRunnersSeeminglyBusy`.

**Comment by mumoshu** [comment]
This should be fixed via #1579. If anyone could double-check if it's really fixed, it would be great if you could build ARC from our current master branch, or try 0.25.0 which will be released soon!
